File: fix-cloudflare-dns.js
#!/usr/bin/env node

// This script will attempt to fix the Cloudflare DNS using the MCP tools
// Since the MCP tools are returning [object Object], we'll try different approaches

const https = require('https');

// Cloudflare API credentials (you'll need to provide these)
const CF_EMAIL = process.env.CF_EMAIL || '';
const CF_API_KEY = process.env.CF_API_KEY || '';
const CF_API_TOKEN = process.env.CF_API_TOKEN || '';

// DNS records we need to set
const DNS_RECORDS = {
  root: {
    type: 'CNAME',
    name: 'generative-engine.org',
    content: 'endothelial-tickseed-ghfs1ljkfattred4hzt4my28.herokudns.com',
    proxied: false
  },
  www: {
    type: 'CNAME', 
    name: 'www',
    content: 'polar-reindeer-y0g7jlud4jpcv07h7675juuy.herokudns.com',
    proxied: false
  }
};

async function makeAPICall(path, method = 'GET', data = null) {
  return new Promise((resolve, reject) => {
    const options = {
      hostname: 'api.cloudflare.com',
      port: 443,
      path: `/client/v4${path}`,
      method: method,
      headers: {
        'Content-Type': 'application/json'
      }
    };

    // Use token if available, otherwise use email/key
    if (CF_API_TOKEN) {
      options.headers['Authorization'] = `Bearer ${CF_API_TOKEN}`;
    } else if (CF_EMAIL && CF_API_KEY) {
      options.headers['X-Auth-Email'] = CF_EMAIL;
      options.headers['X-Auth-Key'] = CF_API_KEY;
    } else {
      reject(new Error('No Cloudflare API credentials found'));
      return;
    }

    const req = https.request(options, (res) => {
      let body = '';
      res.on('data', (chunk) => body += chunk);
      res.on('end', () => {
        try {
          const response = JSON.parse(body);
          if (response.success) {
            resolve(response.result);
          } else {
            reject(new Error(JSON.stringify(response.errors)));
          }
        } catch (e) {
          reject(e);
        }
      });
    });

    req.on('error', reject);
    
    if (data) {
      req.write(JSON.stringify(data));
    }
    
    req.end();
  });
}

async function fixDNS() {
  try {
    console.log('🔍 Finding zone ID for generative-engine.org...');
    
    // Get zone ID
    const zones = await makeAPICall('/zones?name=generative-engine.org');
    if (!zones || zones.length === 0) {
      throw new Error('Zone not found');
    }
    
    const zoneId = zones[0].id;
    console.log(`✅ Found zone ID: ${zoneId}`);
    
    // Get existing DNS records
    console.log('\n📋 Getting existing DNS records...');
    const records = await makeAPICall(`/zones/${zoneId}/dns_records`);
    
    // Find and update root domain record
    const rootRecord = records.find(r => 
      r.name === 'generative-engine.org' && 
      (r.type === 'CNAME' || r.type === 'A')
    );
    
    if (rootRecord) {
      console.log(`\n🔄 Updating root domain record (ID: ${rootRecord.id})...`);
      await makeAPICall(
        `/zones/${zoneId}/dns_records/${rootRecord.id}`,
        'PUT',
        DNS_RECORDS.root
      );
      console.log('✅ Root domain updated');
    } else {
      console.log('\n➕ Creating root domain record...');
      await makeAPICall(
        `/zones/${zoneId}/dns_records`,
        'POST',
        DNS_RECORDS.root
      );
      console.log('✅ Root domain created');
    }
    
    // Find and update www record
    const wwwRecord = records.find(r => 
      r.name === 'www.generative-engine.org' && 
      (r.type === 'CNAME' || r.type === 'A')
    );
    
    if (wwwRecord) {
      console.log(`\n🔄 Updating www subdomain record (ID: ${wwwRecord.id})...`);
      await makeAPICall(
        `/zones/${zoneId}/dns_records/${wwwRecord.id}`,
        'PUT',
        DNS_RECORDS.www
      );
      console.log('✅ WWW subdomain updated');
    } else {
      console.log('\n➕ Creating www subdomain record...');
      await makeAPICall(
        `/zones/${zoneId}/dns_records`,
        'POST',
        DNS_RECORDS.www
      );
      console.log('✅ WWW subdomain created');
    }
    
    console.log('\n🎉 DNS configuration complete!');
    console.log('⏱️  Please wait 5-10 minutes for DNS propagation.');
    console.log('\n📝 Summary:');
    console.log('  - Root domain → Heroku DNS (not proxied)');
    console.log('  - WWW subdomain → Heroku DNS (not proxied)');
    
  } catch (error) {
    console.error('\n❌ Error:', error.message);
    console.log('\n💡 To fix this manually:');
    console.log('1. Set CF_API_TOKEN environment variable with your Cloudflare API token');
    console.log('2. Run this script again: node fix-cloudflare-dns.js');
    console.log('\nOr use the Cloudflare dashboard to update DNS records manually.');
  }
}

// Check if we have credentials
if (!CF_API_TOKEN && (!CF_EMAIL || !CF_API_KEY)) {
  console.log('⚠️  No Cloudflare API credentials found in environment variables.');
  console.log('\nPlease set one of the following:');
  console.log('  - CF_API_TOKEN (recommended)');
  console.log('  - CF_EMAIL and CF_API_KEY');
  console.log('\nExample:');
  console.log('  export CF_API_TOKEN="your-api-token"');
  console.log('  node fix-cloudflare-dns.js');
  process.exit(1);
}

fixDNS();

File: fix_cloudflare_dns.py
#!/usr/bin/env python3
"""
Fix Cloudflare DNS for generative-engine.org
This script updates the DNS records to point to the correct Heroku targets.
"""

import os
import sys
import json
import requests
from typing import Dict, Optional

# DNS Configuration
DNS_CONFIG = {
    'root': {
        'type': 'CNAME',
        'name': 'generative-engine.org',
        'content': 'endothelial-tickseed-ghfs1ljkfattred4hzt4my28.herokudns.com',
        'proxied': False,
        'ttl': 1  # Auto
    },
    'www': {
        'type': 'CNAME',
        'name': 'www',
        'content': 'polar-reindeer-y0g7jlud4jpcv07h7675juuy.herokudns.com',
        'proxied': False,
        'ttl': 1  # Auto
    }
}

class CloudflareAPI:
    def __init__(self, api_token: str = None):
        self.api_token = api_token or os.environ.get('CF_API_TOKEN')
        if not self.api_token:
            print("❌ No Cloudflare API token found!")
            print("\nTo fix the DNS, you need to provide a Cloudflare API token.")
            print("\nOption 1: Set environment variable")
            print("  export CF_API_TOKEN='your-token-here'")
            print("  python3 fix_cloudflare_dns.py")
            print("\nOption 2: Create an API token at:")
            print("  https://dash.cloudflare.com/profile/api-tokens")
            print("  Permissions needed: Zone:DNS:Edit")
            sys.exit(1)
        
        self.headers = {
            'Authorization': f'Bearer {self.api_token}',
            'Content-Type': 'application/json'
        }
        self.base_url = 'https://api.cloudflare.com/client/v4'
    
    def get_zone_id(self, domain: str) -> Optional[str]:
        """Get the zone ID for a domain"""
        url = f"{self.base_url}/zones?name={domain}"
        response = requests.get(url, headers=self.headers)
        
        if response.status_code == 200:
            data = response.json()
            if data['success'] and data['result']:
                return data['result'][0]['id']
        return None
    
    def list_dns_records(self, zone_id: str) -> list:
        """List all DNS records for a zone"""
        url = f"{self.base_url}/zones/{zone_id}/dns_records"
        response = requests.get(url, headers=self.headers)
        
        if response.status_code == 200:
            data = response.json()
            if data['success']:
                return data['result']
        return []
    
    def update_dns_record(self, zone_id: str, record_id: str, record_data: Dict) -> bool:
        """Update an existing DNS record"""
        url = f"{self.base_url}/zones/{zone_id}/dns_records/{record_id}"
        response = requests.put(url, headers=self.headers, json=record_data)
        return response.status_code == 200 and response.json().get('success', False)
    
    def create_dns_record(self, zone_id: str, record_data: Dict) -> bool:
        """Create a new DNS record"""
        url = f"{self.base_url}/zones/{zone_id}/dns_records"
        response = requests.post(url, headers=self.headers, json=record_data)
        return response.status_code == 200 and response.json().get('success', False)
    
    def delete_dns_record(self, zone_id: str, record_id: str) -> bool:
        """Delete a DNS record"""
        url = f"{self.base_url}/zones/{zone_id}/dns_records/{record_id}"
        response = requests.delete(url, headers=self.headers)
        return response.status_code == 200 and response.json().get('success', False)

def main():
    print("🚀 Cloudflare DNS Fix for generative-engine.org")
    print("=" * 50)
    
    # Initialize API
    api = CloudflareAPI()
    
    # Get zone ID
    print("\n🔍 Finding zone ID...")
    zone_id = api.get_zone_id('generative-engine.org')
    
    if not zone_id:
        print("❌ Could not find zone for generative-engine.org")
        print("   Make sure the domain is in your Cloudflare account")
        sys.exit(1)
    
    print(f"✅ Found zone ID: {zone_id}")
    
    # Get existing records
    print("\n📋 Fetching existing DNS records...")
    records = api.list_dns_records(zone_id)
    
    # Process root domain
    print("\n🔧 Processing root domain (generative-engine.org)...")
    root_records = [r for r in records if r['name'] == 'generative-engine.org' and r['type'] in ['A', 'CNAME']]
    
    if root_records:
        # Delete A records if they exist (we need CNAME)
        for record in root_records:
            if record['type'] == 'A':
                print(f"  🗑️  Deleting A record: {record['content']}")
                api.delete_dns_record(zone_id, record['id'])
            elif record['type'] == 'CNAME':
                if record['content'] != DNS_CONFIG['root']['content'] or record['proxied'] != False:
                    print(f"  🔄 Updating CNAME record...")
                    api.update_dns_record(zone_id, record['id'], DNS_CONFIG['root'])
                    print(f"  ✅ Updated: {DNS_CONFIG['root']['content']} (DNS Only)")
                else:
                    print(f"  ✅ Already correct: {record['content']} (DNS Only)")
    else:
        print(f"  ➕ Creating CNAME record...")
        api.create_dns_record(zone_id, DNS_CONFIG['root'])
        print(f"  ✅ Created: {DNS_CONFIG['root']['content']} (DNS Only)")
    
    # Process www subdomain
    print("\n🔧 Processing www subdomain...")
    www_records = [r for r in records if r['name'] == 'www.generative-engine.org' and r['type'] in ['A', 'CNAME']]
    
    if www_records:
        for record in www_records:
            if record['type'] == 'A':
                print(f"  🗑️  Deleting A record: {record['content']}")
                api.delete_dns_record(zone_id, record['id'])
            elif record['type'] == 'CNAME':
                if record['content'] != DNS_CONFIG['www']['content'] or record['proxied'] != False:
                    print(f"  🔄 Updating CNAME record...")
                    api.update_dns_record(zone_id, record['id'], DNS_CONFIG['www'])
                    print(f"  ✅ Updated: {DNS_CONFIG['www']['content']} (DNS Only)")
                else:
                    print(f"  ✅ Already correct: {record['content']} (DNS Only)")
    else:
        print(f"  ➕ Creating CNAME record...")
        api.create_dns_record(zone_id, DNS_CONFIG['www'])
        print(f"  ✅ Created: {DNS_CONFIG['www']['content']} (DNS Only)")
    
    print("\n" + "=" * 50)
    print("🎉 DNS configuration complete!")
    print("\n📝 Summary:")
    print("  ✅ Root domain → Heroku DNS (not proxied)")
    print("  ✅ WWW subdomain → Heroku DNS (not proxied)")
    print("\n⏱️  Please wait 5-10 minutes for DNS propagation")
    print("\n🔍 Test with:")
    print("  curl -I https://generative-engine.org")
    print("  curl -I https://www.generative-engine.org")

if __name__ == "__main__":
    main()